dbcrossbarlib/src/drivers/redshift/write_remote_data.rs

```diff
@@ -56,19 +61,156 @@ pub(crate) async fn write_remote_data_helper(
+    }
+
+    Ok(vec![dest.boxed()])
+}
+
+/// Copy data from S3 into a RedShift table.
+async fn copy_in(
+    ctx: &Context,
+    client: &Client,
+    source_s3_url: &Url,
+    dest_table: &TableName,
+    to_args: &DriverArguments,
+) -> Result<()> {
+    debug!(
+        ctx.log(),
+        "Copying into {} from {}",
+        dest_table.unquoted(),
+        source_s3_url.as_str(),
+    );
     let copy_sql = format!(
         "COPY {dest} FROM {source}\n{credentials}FORMAT CSV\nIGNOREHEADER 1\nDATEFORMAT 'auto'\nTIMEFORMAT 'auto'",
-        dest = TableName(table_name),
-        source = pg_quote(source_url.as_str()), // `$1` doesn't work here.
+        dest = dest_table.quoted(),
+        source = pg_quote(source_s3_url.as_str()), // `$1` doesn't work here.
         credentials = credentials_sql(to_args)?,
     );
     let copy_stmt = client.prepare(&copy_sql).await?;
     client.execute(&copy_stmt, &[]).await.with_context(|_| {
-        format!("error copying {} from {}", pg_create_table.name, source_url)
+        format!(
+            "error copying to {} from {}",
+            dest_table.quoted(),
+            source_s3_url
+        )
     })?;
-    Ok(vec![dest.boxed()])
+    Ok(())
+}
+
+/// Upsert from `temp_table` into `dest_table`, using the columns `upsert_keys`.
+async fn upsert_from_temp_table(
+    ctx: &Context,
+    client: &mut Client,
+    temp_table: &PgCreateTable,
+    dest_table: &PgCreateTable,
+    upsert_keys: &[String],
+) -> Result<()> {
+    let transaction = client.transaction().await?;
+
+    let upsert_sql = upsert_sql(&temp_table, &dest_table, upsert_keys)?;
+    for (idx, sql) in upsert_sql.iter().enumerate() {
+        debug!(
+            ctx.log(),
+            "upsert SQL ({}/{}): {}",
+            idx + 1,
+            upsert_sql.len(),
+            sql,
+        );
+        transaction.execute(&sql[..], &[]).await.with_context(|_| {
+            format!(
+                "error upserting into {} from {}",
+                dest_table.name.quoted(),
+                temp_table.name.quoted(),
+            )
+        })?;
+    }
+
+    debug!(ctx.log(), "commiting upsert");
+    transaction.commit().await?;
+    Ok(())
+}
+
+/// Generate the SQL needed to perform an upsert.
+///
+/// This will destructively modify and then delete `temp_table`.
+fn upsert_sql(
+    temp_table: &PgCreateTable,
+    dest_table: &PgCreateTable,
+    upsert_keys: &[String],
+) -> Result<Vec<String>> {
+    let value_cols = columns_to_update_for_upsert(dest_table, upsert_keys)?;
+    let dest_table_name = dest_table.name.quoted();
+    let temp_table_name = temp_table.name.quoted();
+    let keys_match = upsert_keys
+        .iter()
+        .map(|k| {
+            format!(
+                "{dest_table}.{name} = {temp_table}.{name}",
+                name = Ident(&k),
+                dest_table = dest_table_name,
+                temp_table = temp_table.name.quoted(),
```

Comment:
i don't get how this relates to the localvar above

Reply:
@seamusabshere If I write:

```rust
    let upsert_sql = upsert_sql(&temp_table, &dest_table, upsert_keys)?;
```

The local variable `upsert_sql` "shadows" the function `upsert_sql`. This is just a minor unfortunate clash of otherwise consistent naming conventions. Between Rust's type system, the rich editor support, and the fact that "shadowing" is reasonably idiomatic in Rust, I decided to accept this and to keep internal naming more consistent with other drivers.